csi/controller.go

```diff
@@ -980,6 +1016,88 @@ func (s *OsdCsiServer) CreateSnapshot(
 		},
 	}, nil
 }
+func (s *OsdCsiServer) getCloudBackupClient(ctx context.Context) (api.OpenStorageCloudBackupClient, error) {
+	// Get grpc connection
+	conn, err := s.getRemoteConn(ctx)
+	if err != nil {
+		return nil, status.Errorf(
+			codes.Unavailable,
+			"Unable to connect to SDK server: %v", err)
+	}
+	return s.cloudBackupClient(conn), nil
+}
+
+func (s *OsdCsiServer) createCloudBackup(
+	ctx context.Context,
+	req *csi.CreateSnapshotRequest,
+) (*csi.CreateSnapshotResponse, error) {
+	cloudBackupClient, err := s.getCloudBackupClient(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	// Get any labels passed in by the CO
+	_, locator, _, err := s.specHandler.SpecFromOpts(req.GetParameters())
+	if err != nil {
+		return nil, status.Errorf(codes.InvalidArgument, "Unable to get parameters: %v", err)
+	}
+
+	backupID := req.GetName()
+	// Create snapshot
+	_, err = cloudBackupClient.Create(ctx, &api.SdkCloudBackupCreateRequest{
+		VolumeId: req.GetSourceVolumeId(),
+		TaskId:   backupID,
+		Labels:   locator.GetVolumeLabels(),
+	})
+
+	if err != nil {
+		return nil, status.Errorf(codes.Aborted, "Failed to create cloud snapshot: %v", err)
+	}
+
+	var isBackupReady bool
+	var backupStatus *api.SdkCloudBackupStatusResponse
+
+	// Check if snapshot has been created but is in error state
+	backupStatus, errFindFailed := cloudBackupClient.Status(ctx, &api.SdkCloudBackupStatusRequest{
+		VolumeId: req.GetSourceVolumeId(),
+		TaskId:   backupID,
+	})
+	if errFindFailed != nil {
+		return nil, status.Errorf(codes.Aborted, "Failed to create cloud snapshot: %v", err)
+	}
+	isBackupReady = backupStatus.Statuses[backupID].Status == api.SdkCloudBackupStatusType_SdkCloudBackupStatusTypeDone ||
+		backupStatus.Statuses[backupID].Status == api.SdkCloudBackupStatusType_SdkCloudBackupStatusTypeActive
+	if errFindFailed == nil && !isBackupReady {
+
+		logrus.Infof("cleaned up failed cloud snapshot %v in order to retry snapshot create", backupID)
+		// If snapshot was created but has errors, cleanup and re-create it on the next iteration.
+		_, errCleanupSnap := cloudBackupClient.Delete(ctx, &api.SdkCloudBackupDeleteRequest{
+			BackupId: backupID,
+		})
+		if errCleanupSnap != nil {
+			return nil, status.Errorf(codes.Aborted, "Snapshot create failed and unable to cleanup failed cloud snapshot %s: %v",
+				backupID,
+				errCleanupSnap)
+		}
+	}
+
+	snapSize, errSizeFailed := cloudBackupClient.Size(ctx, &api.SdkCloudBackupSizeRequest{
+		BackupId: backupID,
+	})
+	if errSizeFailed != nil {
+		return nil, status.Errorf(codes.Aborted, "Failed to get cloud snapshot size: %v", err)
+	}
+
+	return &csi.CreateSnapshotResponse{
+		Snapshot: &csi.Snapshot{
+			SizeBytes:      int64(snapSize.GetSize()),
+			SnapshotId:     backupID,
+			SourceVolumeId: req.GetSourceVolumeId(),
+			CreationTime:   backupStatus.Statuses[backupID].StartTime,
+			ReadyToUse:     isBackupReady,
```

Comment:
Don't we need to change this flag to `true` at some point in time? 
I think previously our capability was only for LocalSnapshots which were instantaneous. We might have to change that since cloud backups take some time. 

Check this comment on ListSnapshots
```
// ListSnapshots lists all snapshots in a cluster.
// This is mainly implemented for Nomad, as Kubernetes will not call
// list snapshots for drivers which have synchronous snapshot creation.
// This is because ReadyToUse is set to true immediately upon on our snapshot creation.
```
cc : @ggriffiths

Reply:
I see an implementation in list snapshot where the snapshot status is fetched first before returning it to the caller of the API.
for cloud backups, we could first set it to a pending state and fetch the status of the snaps whenever the list APIs are invoked.
I will raise that change in the list snapshot PR

Reply:
But right now if the backup status is Active you will set this to `true`. Before we merge this change we should fix that. Only set isBackupReady to `true` iff status is SdkCloudBackupStatusType_SdkCloudBackupStatusTypeDone